src/System.IO.FileSystem.Watcher.Polling/src/System/IO/PollingWatcher.cs

```diff
@@ -0,0 +1,141 @@
+
+using System.Threading;
+
+namespace System.IO.FileSystem
+{
+    /// <summary>
+    /// PollingWatcher can be used to monitor changes to a file system directory
+    /// </summary>
+    /// <remarks>
+    /// This type is similar to FileSystemWatcher, but unlike FileSystemWatcher it is fully reliable,
+    /// at the cost of some perfromance overhead. 
+    /// Instead of relying on Win32 file notification APIs, it preodically scans the watched directory to discover changes.
+    /// This means that sooner or later it will discover every change.
+    /// FileSystemWatcher's Win32 APIs can drop some events in rare circumstances, which is often acceptable compromise.
+    /// In scenarios where events cannot be missed, PollingWatcher should be used.
+    /// Note: When a watched file is renamed, one or two notifications will be made.
+    /// Note: When no changes are detected, PollingWatcher will not allocate memory on the GC heap.
+    /// </remarks>
+    public class PollingWatcher : IDisposable
+    {
+        Timer _timer;
+        int _pollingIntervalInMilliseconds;
+
+        string _directory;
+        PathToFileStateHashtable _state = new PathToFileStateHashtable(); // stores state of the directory
+        byte _version; // this is used to keep track of removals. // TODO: describe the algorithm
+
+        /// <summary>
+        /// Creates an instance of a watcher
+        /// </summary>
+        /// <param name="rootDirectory">The directory to watch. It does not support UNC paths (yet).</param>
+        /// <param name="pollingIntervalInMilliseconds">Polling interval</param>
+        public PollingWatcher(string rootDirectory, int pollingIntervalInMilliseconds = 1000)
+        {
+            _pollingIntervalInMilliseconds = pollingIntervalInMilliseconds;
+            _directory = @"\\?\" + rootDirectory + @"\*"; // TODO: this needs to be smarter for UNC paths
+            ComputeChangesAndUpdateState(_directory); // captures the initial state
+
+            _timer = new Timer(new TimerCallback(TimerHandler), null, pollingIntervalInMilliseconds, Timeout.Infinite);
+        }
+
+        /// <summary>
+        /// This callback is called when any change (Created, Deleted, Changed) is detected.
+        /// </summary>
+        public Action Changed;
+
+        /// <summary>
+        /// Disposes the timer used for polling.
+        /// </summary>
+        public void Dispose()
+        {
+            _timer.Dispose();
+        }
+
+        private void TimerHandler(object context)
+        {
+            var handler = Changed;
+            if (handler != null)
+            {
+                var changes = ComputeChangesAndUpdateState(_directory);
+                if (!changes.IsEmpty)
+                {
+                    handler();
+                }
+            }
+
+            _timer.Change(_pollingIntervalInMilliseconds, Timeout.Infinite);
+        }
+        private FileChangeList ComputeChangesAndUpdateState(string directory)
```

Comment:
I just ran quick and dirty test on a folder with:
a) 4000 files. One polling cycle takes ~40ms (~150k ticks).
b) 500 files. ~7000 ticks (1 ms)
c) 100 files -> ~700 ticks (approaching 0ms)

The hot code is in the Equal(string, WIN32_FIND_DATAW) routine. It's responsible for ~50% of the samples. I can optimize the routine and in addition, the routine is probably called too much because of hashatable bucket conflicts, which I can also improve by making it sparser.